crates/bevy_ecs/src/component.rs

```diff
@@ -283,22 +284,55 @@ impl ComponentInfo {
 /// one `World` to access the metadata of a `Component` in a different `World` is undefined behavior
 /// and must not be attempted.
 #[derive(Debug, Copy, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
-pub struct ComponentId(usize);
+pub struct ComponentId {
+    /// The actual index of the `ComponentId`. This uniquely identifies the component this id refers
+    /// to.
+    index: u32,
+    /// An optional `Entity`, which is used for relations. If this is a non-relation
+    /// `ComponentId`, the value of this field is `ComponentId::NO_TARGET`.
+    target: Entity,
```

Comment:
Is the `Entity.generation` relevant here? Could we get away with only storing the entity index?

Reply:
The generation was added at Alice's request to prevent collisions until cleanup policies are introduced, which would provide an alternative to prevent collisions. Otherwise this PR would be quite broken on its own.